service/src/LevelZero.cpp

```diff
@@ -523,6 +526,366 @@ namespace geopm
+        if (ze_result == ZE_RESULT_SUCCESS) {
+            unsigned int num_reports = num_metric_values / num_metric;
+
+            for (unsigned int metric_idx = 0; metric_idx < num_metric; metric_idx++)
+            {
+                //TODO: It is possible that simply parsing all the metrics is
+                //      faster than the additional API calls to check the metric
+                //      name.  This should be studied
+                std::vector<zet_metric_handle_t> metric_handle(num_metric);
+                ze_result = zetMetricGet(m_devices.at(l0_device_idx).subdevice.metric_group_handle.at(l0_domain_idx),
+                                         &num_metric, metric_handle.data());
+                check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+                                "LevelZero::" + std::string(__func__) +
+                                ": LevelZero Metric handle acquisition failed",
+                                __LINE__);
+
+                // process metrics
+                zet_metric_properties_t metric_properties;
+                ze_result = zetMetricGetProperties(metric_handle.at(metric_idx), &metric_properties);
+                check_ze_result(ze_result, GEOPM_ERROR_RUNTIME,
+                                "LevelZero::" + std::string(__func__) +
+                                ": LevelZero Metric property acquisition failed",
+                                __LINE__);
+
+                std::string metric_name (metric_properties.name);
+
+                //TODO: check timing impact of only processing metrics supported by IOGroup
+                if(metric_name.compare("XVE_ACTIVE") == 0 ||
+                   metric_name.compare("XVE_STALL") == 0) {
+                    for (unsigned int report_idx = 0; report_idx < num_reports; report_idx++)
+                    {
+                        //This is the actual gathering of the data
+                        zet_typed_value_t data = metric_values.at(report_idx*num_metric + metric_idx);
+                        double data_double = NAN;
+                        switch( data.type )
+                        {
+                        case ZET_VALUE_TYPE_UINT32:
+                            data_double = data.value.ui32;
+                            break;
+                        case ZET_VALUE_TYPE_UINT64:
+                            data_double = data.value.ui64;
+                            break;
+                        case ZET_VALUE_TYPE_FLOAT32:
+                            data_double = data.value.fp32;
+                            break;
+                        case ZET_VALUE_TYPE_FLOAT64:
+                            data_double = data.value.fp64;
+                            break;
+                        case ZET_VALUE_TYPE_BOOL8:
+                            data_double = data.value.ui32;
+                            break;
+                        default:
+                            break;
+                        };
```

Comment:
Does the API provide a helper function to do this? If not, consider adding a helper function to do this. If it's not needed anywhere else it would at least reduce this function's size and make it more readable.

Reply:
The API does not, I've moved it to metric_data_convert()